Upload image

will-weiss opened the pull request.

shachar-langer commented:
I pulled the latest code from both the server and the extension but I couldn't post a tweet successfully. I'm investigating to see if it's a problem in my setup.

shachar-langer commented:
It was a server [issue](https://github.com/morehumaninternet/roar-server/pull/29). Fixed.

will-weiss commented:
Hey @BarbaraPruz ! Took a stab at the rebase and made a few changes that I hope make sense. Let me know if you have any questions, but if we get this merged we can move onto adding support for other image extensions and drag and drop, as discussed.

- I moved the hidden file input into the action bar so that it could be clicked like so `<ActionButton kind="AddImage" onClick={() => imageRef.current!.click()} disabled={takeScreenshotDisabled} />`. That’s a trick I’ve used a few times (like on the [MHI frontend](https://github.com/morehumaninternet/frontend/blob/production/src/components/shared/application-form/file-upload-button.tsx#L28)) that continues to be useful
- Instead of using an `any` as the type of `file` I updated the `UserAction` and other places to read `{ type: 'imageUpload'; payload: { file: File } }`. This way typescript can follow along, especially for something like `readFileUri`
- Rather than persist the `uploadImageFile` we need only listen for the action with `store.on('imageUpload', state => {` and then retrieve the file from `state.mostRecentAction.payload`. This means the responder doesn’t even need to do anything at all, just the listener.
- I use the file’s name as the name rather than inspecting the tab’s URL

shachar-langer commented on src/background/images.ts:
```diff
@@ -49,8 +50,32 @@ export async function takeScreenshot(
     })
   } catch (error) {
-    dispatchBackgroundActions.screenshotCaptureFailure({ error })
+    dispatchBackgroundActions.imageCaptureFailure({ targetId: tab.id, error })
```
I'm not against sending the `targetId` but it doesn't look like we're using it

will-weiss replied:
Let's leave it alone, this will be useful when handling image loading

shachar-langer commented on src/background/responders.ts:
```diff
@@ -135,28 +137,29 @@ export const responders: Responders<Action> = {
       },
     }))
   },
+  imageUpload(state, { file }): Partial<StoreState> {
```
Would we save the file or the file's data in the state? Why are we passing the file here?

will-weiss replied:
It's accessed by the listener, but not used here, so you're right that we can omit these from the function arguments of the responder.

BarbaraPruz replied:
Fixed in latest checkin.

shachar-langer commented on src/background/run.ts:
```diff
@@ -30,18 +30,27 @@ export function run(backgroundWindow: Window, browser: typeof global.browser, ch
   store.on('clickTakeScreenshot', state => {
-    takeScreenshot(ensureActiveFeedbackTarget(state), browser.tabs, store.dispatchers)
+    const target = ensureActiveFeedbackTarget(state)
+    if (addImageDisabled(target)) return
```
If I'm not mistaken, the button will be disabled so this is redundant.

will-weiss replied:
good call, lets' remove those

BarbaraPruz replied:
Fixed in latest checkin.

shachar-langer commented on src/background/run.ts:
```diff
@@ -30,18 +30,27 @@ export function run(backgroundWindow: Window, browser: typeof global.browser, ch
+
+  store.on('imageUpload', state => {
+    const target = ensureActiveFeedbackTarget(state)
+    if (addImageDisabled(target)) return
```
Same comment as above

BarbaraPruz replied:
Fixed in latest checkin
